File: python/ModifiedFib.py
class ModifiedFib(object):
  
    def __init__(self, a, b, n):
        self.a = a
        self.b = b
        self.n = n
        
    def run_fib(self):
        self.run_fib_accum(self.a, self.b, 3)

    def run_fib_accum(self, a_accum, b_accum, i):
        next_val = a_accum + b_accum ** 2
        
        if (i == self.n):
            # Base case
            print str(next_val)
        else:
            self.run_fib_accum(b_accum, next_val, i + 1)

input_string = raw_input().split(' ')

modified_fib = ModifiedFib(
    int(input_string[0]),
    int(input_string[1]),
    int(input_string[2]))

modified_fib.run_fib()

File: java/ModifiedFib.java
import java.io.*;
import java.util.*;
import java.math.BigInteger;

public class Solution {
    
    private static class FibRunner {
        private final BigInteger a;
        private final BigInteger b;
        private final int n;
        
        public FibRunner(int a, int b, int n) {
            this.a = BigInteger.valueOf(new Integer(a).longValue());
            this.b = BigInteger.valueOf(new Integer(b).longValue());
            this.n = n;
        }
        
        public void runFib() {
            runFibAccum(this.a, this.b, 3);
        }
        
        private void runFibAccum(BigInteger aAccum, BigInteger bAccum, int i) {
            BigInteger nextVal = aAccum.add(bAccum.pow(2));
                        
            if (i == this.n) {
                // base case
                System.out.println(nextVal.toString());
            } else {
                this.runFibAccum(bAccum, nextVal, i + 1);
            }
        }
    }

    public static void main(String[] args) {
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution. */
        Scanner sc = new Scanner(System.in);
        
        int a = sc.nextInt();
        int b = sc.nextInt();
        int n = sc.nextInt();
        
        FibRunner fibRunner = new FibRunner(a, b, n);
        fibRunner.runFib();
    }
}